### PR TITLE
Refactor the main script

### PR DESCRIPTION
**Pull request by nfitzen**
I was unsure what you intended with the `community` folder, so maybe this PR isn't in line with that. It certainly isn't a bug fix. I'm happy to move this improved script to the `community` folder if you want to retain authenticity from the video. (That said, I didn't change the algorithm, and I signed any comments I wrote with `-N`.)

The following is a list of stuff I did. Probably not exhaustive.

- Add type hints
- Add optional arguments
- Allow the file to be an importable module for further work (I just separated out the script portion).
- Use command line arguments
- Format with black
- File name uses underscores (necessary for importability)
- Add very simple testing
- Update README accordingly
- Use `break` in while loop

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,213 @@
+# SPDX-License-Identifier: MIT
```
Comment (oliverdunk) [inline .gitignore:1]: Did you notice any files being generated that caused you to add this? I think we should go for a simple `.gitignore` with only files we expect to run in to.

File: README.md [hunk 2]
```diff
@@ -1,12 +1,18 @@
+<!-- SPDX-License-Identifier: MIT -->
```
Comment (oliverdunk) [inline README.md:1]: Do you have an example of other projects that use an SPDX header? I'm not totally against it but it isn't something I've seen before.

File: LICENSE [hunk 3]
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2020 standupmaths
+Copyright (C) 2020-2021 standupmaths and the bubble-cosh contributors
```
Comment (oliverdunk) [inline LICENSE:3]: Good spot, thanks!

File: bubble_cosh.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: MIT
+# Copyright (C) 2020-2021 standupmaths and the bubble-cosh contributors
+
+# The type hints use syntax from Python 3.9+. See also PEP 585 -N
```
Comment (oliverdunk) [inline bubble_cosh.py]: I think we can probably drop the `-N` in comments? You'll still get credit in the blame but the initialing is probably just confusing!

File: bubble_cosh.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+
+def total_area(a, b, d, l):
+    return pi * (a ** 2) * (sinh(l / a) + (l / a))
+
+
+def main():
+    # This is messy since it's internal behavior. argparse really should have
+    # a union operator of some sort. -N
+    custom_format = lambda prog: argparse.ArgumentDefaultsHelpFormatter(prog, width=72)
```
Comment (oliverdunk) [inline bubble_cosh.py:89]: Am I understanding correctly that you did this to allow overriding the width? If so, maybe mentioning that would be a more helpful comment?

**Review by oliverdunk — CHANGES_REQUESTED** [review]
Hey @nfitzen! A big thank you for this. Lots of great tidy-up and probably some better Python (I don't think Matt would mind me saying that 😅). The honest answer is that we're still figuring out which PRs we should and shouldn't accept. Everything in here makes the code easier to use, though, so I think I'm happy with the idea of merging it.

I've left a few comments. Let me know when you've taken a look at those and I can give this another pass.

File: .gitignore (hunk 1 above)
Comment (nfitzen) [reply .gitignore:1]: True. It was mostly the `__pycache__` folder.

File: README.md (hunk 2 above)
Comment (nfitzen) [reply README.md:1]: [Linux](https://www.kernel.org/doc/html/latest/process/license-rules.html#kernel-licensing) is the most notable project.
I know there are a few others, but regardless, it's good for license clarity.

Edit: Fixed link and changed last sentence a bit.

File: bubble_cosh.py (hunk 4 above)
Comment (nfitzen) [reply bubble_cosh.py]: > You'll still get credit in the blame

I didn't do it for the credit; I was just making sure Matt's comments were marked as such (by omission). There's no issue with removing it, so I'm going to do that.

File: bubble_cosh.py (hunk 5 above)
Comment (nfitzen) [reply bubble_cosh.py:89]: I made a mention just now.

**Comment by nfitzen** [comment]
> I've left a few comments. Let me know when you've taken a look at those and I can give this another pass.

@oliverdunk sounds good! I made the requested changes. Can you take a look?

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+# Copyright (C) 2017 Version Control System Fund
+# Copyright (C) 2021 nfitzen
```
Comment (oliverdunk) [inline .gitignore:3]: Where did these come from? I think if we need to add more copyright information, I would rather write our own gitignore from scratch so we don't need anything :)

**Review by oliverdunk — CHANGES_REQUESTED** [review]
Sorry for the slow reply! Left one more comment that would be nice to address before merging.

**Pull request closed without merging**